src/clj/witan/send/validate_model.clj

```diff
@@ -0,0 +1,126 @@
+
+(defn collate-fold [input-path year n-transitions]
+  (let [test-data (load-csv-as-maps (str (temp-dir input-path) "test_" year ".csv"))
+        test-years (->> (map :calendar-year test-data)
+                        (distinct)
+                        (map #(inc (Integer/parseInt %)))
+                        (into [])
+                        (map str))
+        model-state (return-testable-data (str (temp-dir input-path) year "_Output_AY_State.csv") test-years)
+        model-count (return-testable-data (str (temp-dir input-path) year "_Output_Count.csv") test-years)
+        count-results (map #(append-count-with-test % test-data n-transitions) model-count)
+        state-results (map #(append-state-with-test % test-data n-transitions) model-state)]
+    (write-csv (str (temp-dir input-path) "results_" year "_count.csv") count-results)
+    (write-csv (str (temp-dir input-path) "results_" year "_state.csv") state-results)))
+
+(defn validate-fold [input-path year transitions settings]
+  (def temp (temp-dir input-path))
+  (let [train-data (return-fold <= year transitions)
+        n-transitions (count (distinct (map :calendar-year train-data)))
+        test-data (return-fold > year transitions)
+        fold-settings (merge settings {:output? true :override-inputs-path (str input-path "temp/" year "/")})]
+    (copy-input-files input-path year)
+    (write-csv (str (temp-dir input-path) year "/transitions.csv") train-data)
+    (write-csv (str (temp-dir input-path) "test_" year ".csv") test-data)
+    (run-model fold-settings)
+    (move-target-files input-path year)
+    (collate-fold input-path year n-transitions)))
+
+(defn run-validation [input-path settings-map keep-temp-files?]
+  "Function to validate SEND model.
+  input-path is a string with the data directory to validate the model on e.g. demo/
+  settings-map expects a hash map containing run-model parameters to use for validation eg. {:iterations 100}
+  (note that run-model's :output? and :override-inputs-path arguments are both required for validation and are overwritten if provided)
+  keep-temp-files? can be set to false unless debugging etc."
+  (.mkdir (java.io.File. (temp-dir input-path)))
+  (let [transitions (load-csv-as-maps (str "data/" input-path "transitions.csv"))
+        years-to-validate (get-validation-years transitions)]
+    (doseq [year years-to-validate]
+      (validate-fold input-path year transitions settings-map))
+    (->> (map #(load-csv-as-maps (str "data/" input-path "temp/results_" % "_count.csv")) years-to-validate)
+         (flatten)
+         (write-csv (str "validation_result_count_" (str/join (drop-last input-path)) ".csv")))
+    (->> (map #(load-csv-as-maps (str "data/" input-path "temp/results_" % "_state.csv")) years-to-validate)
+         (flatten)
+         (write-csv (str "validation_result_state_" (str/join (drop-last input-path)) ".csv"))))
+  (if (false? keep-temp-files?)
```

Comment:
`(not keep-temp-files?)` is more idiomatic, as it covers `false` and `nil`.

Reply:
changed